File: python/1d-dynamic-programming/word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        cache = {}

        def dfs(i):
            if i > len(s):
                return False
            if i == len(s):
                return True
            if i in cache:
                return cache[i]

            cache[i] = False
            for word in wordDict:
                if i + len(word) > len(s) or s[i:i + len(word)] != word:
                    continue
                cache[i] = cache[i] or dfs(i + len(word))

            return cache[i]

        return dfs(0)

File: c++/1d-dynamic-programming/word-break.cc
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        std::unordered_map<int, int> cache;
        return break_word(0, s, wordDict, cache);
    }

private:
    bool break_word(int i, std::string& s, std::vector<std::string>& dictionary,  std::unordered_map<int, int>& cache) {
        if (i == s.size()) {
            return true;
        }

        if (cache.contains(i)) {
            return cache[i];
        }

        cache[i] = false;
        for (std::string word : dictionary) {
            if (word == s.substr(i, word.size())) {
                cache[i] = cache[i] || break_word(i + word.size(), s, dictionary, cache);
            }
        }
        return cache[i];
    }
};
